### ISSUE TITLE
Improving the WaveDiff optimisation procedure

### ISSUE THREAD

**Issue by CentofantiEze** [tobias-liaudat/wf-SEDs#17]
# Project 1: learning projection

### Current Issue
The model is trained in an alternate way between the parametric and the non-parametric contribution. After a non-parametric (powerful model) optimisation cycle the parametric model (really simple) doesn't have any room to improve its contribution. 

### Idea
The idea is that the non-parametric model leads the parametric one to a better solution instead of just correcting its errors. In this way after performing a non-parametric optimisation cycle we let the parametric part to evolve to a better solution. 

Once the non-parametric model is optimised, *transfer* its WFE projection on the first `n_zernikes` to the parametric model coefficients. Do this parameter transfer after each non-parametric optimisation cycle. 

## To do

- [x] Solve [issue 25](https://github.com/tobias-liaudat/wf-psf/issues/25) form the wf-psf repo.
- [x] Analyse the results from #12 , specially the training of the models with the different optimisation strategies.

# Project 2: refining the solution with a joint optimisation

TBD... An issue to tackle, handle a joint optimisation of a different set of parameters with different (an order of magnitude) learning rates

**Comment by tobias-liaudat** [comment]
## Validate project 1

We want to define experiences to see if the new optimisation procedure is helping to learn more information or not.

We would like to see if we can learn more in the case the data is closer to the observations (dataset n1), and in the case when we have a high number of stars (dataset n2).

### Datasets 

Dataset n1
- Parameters: `WFE_res(GT)= 256`, `n_bins(GT)=350`, `n_stars=2000`,
- Inputs: Possibility to use `n_bins=8, 16, 32`

Dataset n2
- Parameters: `WFE_res(GT)= 128`, `n_bins(GT)=8`, `n_stars=4000`, (in fact `500<=n_stars<=10k`)


### Tests


- [ ] `Fiducial model` (without new optimisation strategy). `model =poly`, `d_max_nonparam=5`, `n_zernikes =15` and `d_max=2`. Same `cycles=2`, `learning rate` and `n_epochs` as in the paper.
- [ ] `Model n1`: Fiducial projecting using `2` cycles without projecting the last cycle.
- [ ] `Model n2`: Fiducial projecting using 2 cycles without projecting the last cycle.

**Comment by tobias-liaudat** [comment]
# More work

Repeat the following trainings using the new learning rate strategy. Lineal descent between `0.1`and `0.06`.

- [ ] `Model n2 new_lr`: Fiducial **projecting** using `5 cycles` without projecting the last cycle. (25)
- [ ] `Model n2 bis new_lr`: Fiducial **without** projecting using `5 cycles`. (25)
- [ ] `Model n3 new_lr`: Fiducial **projecting** using `9 cycles` without projecting the last cycle. (13)
- [ ] `Model n3 bis new_lr`: Fiducial **without** projecting using `9 cycles`. (13)

Sanity check, repeat the following training:
- [ ] `Fiducial model` (without new optimisation strategy). `model =poly`, `d_max_nonparam=5`, `n_zernikes =15` and `d_max=2`. Same `cycles=2`, `learning rate` and `n_epochs` as in the paper.
- [ ] `Model n1`: Fiducial **projecting** using `2 cycles` without projecting the last cycle. (100)

**Comment by CentofantiEze** [comment]
# Get better result estimations

Repeat the following trainings 5 times each to be able to take the median or average value:

- [x] `Fiducial model` (without new optimisation strategy). `model =poly`, `d_max_nonparam=5`, `n_zernikes =15` and `d_max=2`. Same `cycles=2`, `learning rate` and `n_epochs` as in the paper.
- [x] `Model n1`: Fiducial **projecting** using `2 cycles` without projecting the last cycle. (100)
- [x] `Model n2 new_lr`: Fiducial **projecting** using `5 cycles` without projecting the last cycle. (25)
- [x] `Model n2 bis new_lr`: Fiducial **without** projecting using `5 cycles`. (25)
- [x] `Model n3 new_lr`: Fiducial **projecting** using `9 cycles` without projecting the last cycle. (13)
- [x] `Model n3 bis new_lr`: Fiducial **without** projecting using `9 cycles`. (13)

1 cycle tests:
- [x] `Model n0` (without new optimisation strategy) using`cycles=1`, `l_rate_param = 0.008`, `n_epochs_param=30`, `l_rate_non_param = 0.075` and`n_epochs_non_param=150`.
- [x] `Model n1`: Fiducial **projecting** using `1 cycles` without projecting the last cycle and using the learning rates as `model n0`. (150)


Document results

- [x] Make a table or plot showing the results over the five optimisation routines using the mean and the median.
- [x] Document to overleaf notes.

**Comment by tobias-liaudat** [comment]
Missing : **Project 2: refining the solution with a joint optimisation**
